# Assignment 1

This assignment will involve the creation of a spellchecking system and an evaluation of its performance. You may use the code snippets provided in Python for completing this or you may use the programming language or environment of your choice

Please start by downloading the corpus `holbrook.txt` from Blackboard

The file consists of lines of text, with one sentence per line. Errors in the line are marked with a `|` as follows

    My siter|sister go|goes to Tonbury .
    
In this case the word 'siter' was corrected to 'sister' and the word 'go' was corrected to 'goes'.

In some places in the corpus two words maybe corrected to a single word or one word to a multiple words. This is denoted in the data using underscores e.g.,

    My Mum goes out some_times|sometimes .
    
For the purpose of this assignment you do not need to separate these words, but instead you may treat them like a single token.

*Note: you may use any functions from NLTK to complete the assignment. It should not be necessary to use other libraries and so please consult with us if your solution involves any other external library. If you use any function from NLTK in Task 6 please include a brief description of this function and how it contributes to your solution.*



### Name - TAPAN AUTI

### Class  - MSc Artificial Intelligence (1MAI1)

### Student ID - 20231499

## Task 1 (10 Marks)

Write a parser that can read all the lines of the file `holbrook.txt` and print out for each line the original (misspelled) text, the corrected text and the indexes of any changes. The indexes refers to the index of the words in the sentence. In the example given, there is only an error in the 10th word and so the list of indexes is [9]. It is not necessary to analyze where the error occurs inside the word.

Then split your data into a test set of 100 lines and a training set.

In [1]:
import nltk
#storing all lines from text holbrook
lines = []  
with open('holbrook.txt', 'r',encoding='utf-8') as f:
    lines = f.readlines()
#storing dictionary values
data = []

#the following code stores the coorected sentence in corrected list and original sentence in original list and counts the indexes of changes.
for line in lines:
    original = []
    corrected = []
    indexes = []
    count = 0
    l = line.split()
    for word in l:       
        if '|' in word:
            var = word.split('|')
            original.append(var[0])
            corrected.append(var[1])
            indexes.append(count)
        else:
            original.append(word)
            corrected.append(word)
        count += 1
    data.append({'original':original, 'corrected':corrected, 'indexes':indexes})

print(data[2])    
    

# Write your code here

#assert(data[2] == {
#    'original': ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'siter', '.'], 
#    'corrected': ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'sister', '.'], 
#    'indexes': [9]
#})

{'original': ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'siter', '.'], 'corrected': ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'sister', '.'], 'indexes': [9]}


The counts and assertions given in the following sections are based on splitting the training and test set as follows

In [ ]:
#spilting the original and the test data into corrected and original 
test = data[:100]
train = data[100:]
train_corrected = [i['corrected'] for i in train]
train_original = [i['original'] for i in train]
test_corrected = [i['corrected'] for i in test]
test_original = [i['original'] for i in test]

## **Task 2** (10 Marks): 
Calculate the frequency (number of occurrences), *ignoring case*, of all words and their unigram probability from the corrected *training* sentences.

*Hint: use `Counter` to implement this so it may be called many times*

In [4]:
from collections import Counter

import itertools

#this function tokenises the corrected sentences from the training data
def seperation():
    test = data[:100]
    train = data[100:]
    train_corrected = [i['corrected'] for i in train]
    train_tokenize = (list(itertools.chain.from_iterable(train_corrected)))
    return train_tokenize

# This function counts the unigram frequency
def unigram(word):
    train_tokenize = seperation()
    counter = Counter(train_tokenize)
    ans = counter[f"{word}"]
    return ans
     
#Tiis functions counts the probability of the unigram
def prob(word):
    train_tokenize = seperation()
    length = len(train_tokenize)
    Prob = unigram(word)/length
    return Prob

def main():
    u = unigram('me')

    p = prob('me')

    return u,p

main()



#If you want to get all words count then use the below code--

# train_tokenise = list(set(seperation()))
# counter = Counter(train_tokenize)
# for i in train_tokenise: 
#     ans = counter[f"{i}"]
#     print(i,ans)


# Test your code with the following
#assert(unigram("me")==87)

, 'moved': 2, 'HE': 2, 'America': 2, 'Paul': 2, 'California': 2, 'flames': 2, 'loaded': 2, 'turned': 2, 'america': 2, 'York': 2, 'dear': 2, 'worry': 2, 'england': 2, 'daughter': 2, 'heart': 2, 'awful': 2, 'everything': 2, 'weekends': 2, 'kinds': 2, 'boss': 2, 'owns': 2, 'cow': 2, 'loads': 2, 'lame': 2, 'feed': 2, 'bran': 2, "couldn't": 2, "isn't": 2, 'gate': 2, 'cart': 2, 'ought': 2, 'seventeen': 2, 'flowers': 2, 'hallo': 2, 'Harison': 2, 'shine': 2, 'join': 2, 'beginning': 2, 'mole': 2, 'built': 2, 'summer': 2, 'die': 2, 'frog': 2, 'spawn': 2, 'cool': 2, 'bigger': 2, 'eggs': 2, 'jar': 2, 'weed': 2, 'tip': 2, 'rough': 2, 'read': 2, 'curtains': 2, 'explore': 2, 'rowing': 2, 'skeleton': 2, 'disappeared': 2, 'wondered': 2, 'plan': 2, 'ourselves': 2, 'treasure': 2, 'creepy': 2, 'tower': 2, 'finger': 2, 'sharp': 2, 'washed': 2, 'somewhere': 2, 'ask': 2, 'washing': 2, 'Dead': 2, 'Kitty': 2, 'Forged': 2, 'Note': 2, 'Be': 2, 'PLENTY': 2, "Frogmen's": 2, 'rex': 2, 'Jacket': 2, 'alone': 2, 'bus'

TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'

## **Task 3** (15 Marks): 
[Edit distance](https://en.wikipedia.org/wiki/Edit_distance) is a method that calculates how similar two strings are to one another by counting the minimum number of operations required to transform one string into the other. There is a built-in implementation in NLTK that works as follows:


In [40]:
from nltk.metrics.distance import edit_distance

# Edit distance returns the number of changes to transform one word to another
print(edit_distance("hello", "hi"))

4


Write a function that calculates all words with *minimal* edit distance to the misspelled word. You should do this as follows

1. Collect the set of all unique tokens in `train`
2. Find the minimal edit distance, that is the lowest value for the function `edit_distance` between `token` and a word in `train`
3. Output all unique words in `train` that have this same (minimal) `edit_distance` value

*Do not implement edit distance, use the built-in NLTK function `edit_distance`*

In [41]:

def get_candidates(token):
    train_tokenize = seperation()
    unique = set(train_tokenize) #gets unique words from the sentences
    dis=[]
    for i in unique:
        dis.append([edit_distance(i,token)])
        # calculates the distance for each token for the words in  training 

    distance = dict(zip(unique,dis))# stores the words and its distance
    sorted_dict = {k: v for k, v in sorted(distance.items(), key=lambda item: item[1])}
    values = min(sorted_dict.values())
    result = [key for key in sorted_dict if sorted_dict[key]==values]
    return result  # gets the words with same distance value
get_candidates('minde')
        
# Test your code as follows
#assert get_candidates("minde") == ['mind', 'mine']

['mind', 'mine']

## Task 4 (15 Marks):

Write a function that takes a (misspelled) sentence and returns the corrected version of that sentence. The system should scan the sentence for words that are not in the dictionary (set of unique words in the training set) and for each word that is not in the dictionary choose a word in the dictionary that has minimal edit distance and has the highest *unigram probability*. 

*Your solution to this should involve `get_candidates`*


In [5]:
def correct(sentence):
    corrected = []
    similar =[]
   
    for i in sentence:
        similar.append(get_candidates(i)) #getting the values with similar edit distances
    #this loop calculates the highest probability between the same values  and returns the word
    for j in similar:
        if len(j)!=1:
            for i in j:
                low = {}
                low[i] = prob(i)
            values = max(low.values())
            result = [key for key in low if low[key]==values]
            corrected.extend(result)
        else:
            corrected.extend(j)
     
    return corrected
correct(["this","whitr","cat"])

#assert(correct(["this","whitr","cat"]) = ['this','white','cat'])   

['this', 'white', 'cat']

## **Task 5** (10 Marks): 
Using the test corpus evaluate the *accuracy* of your method, i.e., how many words from your system's output match the corrected sentence (you should count words that are already spelled correctly and not changed by the system).

In [6]:
from difflib import SequenceMatcher # returns the ratio of similarity
def accuracy():
    test = data[:100]
    algo = []
    acc = []
    test_corrected = [i['corrected'] for i in test]
    test_original = [i['original'] for i in test]
    # this loop calculates the accuracy of the algorithm.
    for i in range(0,len(test_original)):
        algo.append(correct(test_original[i]))
        acc.append(SequenceMatcher(None,test_corrected[i],algo[i]).ratio())
    total_accuracy = ((sum(acc)/len(acc))*100)
    return total_accuracy
accuracy()


79.26613818518715

## **Task 6 (35 Marks):**

Consider a modification to your algorithm that would improve the accuracy of the algorithm developed in Task 3 and 4

* You may resources beyond those provided here.
* You must **not use the test data** in this task.
* Provide a short text describing what you intend to do and why. 
* Full marks for this section may be obtained without an implementation, but an implementation is preferred.
* Your implementation should not consist of more than 50 lines of code

Please note this task is marked according to: demonstration of knowledge from the lectures (10), originality and appropriateness of solution (10), completeness of description (10) and technical correctness (5)


## Solution

### Consider the following example for first three solutions:
    ['NIGEL','THRUSH', 'February', '48'] -->  ['JILL', 'JUST', 'Bernard', '40']
1. Named Entitiy - (Implemented)
    In the algorithm the named entities such as names and organizational names were also condidered as a part of change which was not to be implemented like that, ex- 'NIGEL' was coverted to 'JILL'  So all named entity should be handled correctly and not changed.

2. Numerical Data -(Implemented)
    The numerical data was also changed so this should also be handled.ex - 48 was changed to 40which was inappropriate

3. Dates - 
    For months like 'February' it was changed to 'Bernard' so this should be handled properly.

4. Increase in DataSet - 
    The training data was very less if it would have beem more exhaustive then it might have added the training list ,for correction in turn giving more accuracy.

5. Morphology - 
    There may be cases where the word is not in its reduced form so we can use stemming and lemmatization of such words, basically to have normalized corpus.

6. Smoothing - 
    We can use Language Model Smoothing to reduce zero probabilities

7. Additional Methods - 
    We can further implement Hidden Markov Models to increase accuracy, it will prove to be robust parser in this case.  (
    

In [ ]:
import nltk
from nltk import pos_tag
def correct_tag(sentence):
    corrected = []
    similar =[]

    tags = ['NNP','CD']
    pos = pos_tag(sentence)  # pos tagging   

   
    for i in pos:
        if i[1] in tags:
            similar.append([i[0]])# checking for proper noun and integers
        else:
            similar.append(get_candidates(i[0])) #getting the values with similar edit distances
    #this loop calculates the highest probability between the same values  and returns the word
    for j in similar:
        if len(j)!=1:
            for i in j:
                low = {}
                low[i] = prob(i)
            values = max(low.values())
            result = [key for key in low if low[key]==values]
            corrected.extend(result)
        else:
            corrected.extend(j)
     
    return corrected

## **Task 7 (5 Marks):**

Repeat the evaluation (as in Task 5) of your new algorithm and show that it outperforms the algorithm from Task 3 and 4

In [52]:
from difflib import SequenceMatcher # returns the ratio of similarity
def accuracy():
    test = data[:100]
    algo = []
    acc = []
    test_corrected = [i['corrected'] for i in test]
    test_original = [i['original'] for i in test]
    # this loop calculates the accuracy of the algorithm.
    for i in range(0,len(test_original)):
        algo.append(correct_tag(test_original[i]))
        acc.append(SequenceMatcher(None,test_corrected[i],algo[i]).ratio())
    total_accuracy = ((sum(acc)/len(acc))*100)
    return total_accuracy
accuracy()

87.18168894442178

We can see that in previous implementation we were getting 79 % accuracy but after handling of the proper nouns and integer values there was a significant chnage in the accuracy, it increased by almost 8 percent going to 87 %.